lang/LangSource/PyrSlot.h

```diff
@@ -27,16 +27,398 @@ A PyrSlot is an 8-byte value which is either a double precision float or a
-#    include "PyrSlot32.h"
+#    define POINTER_NEEDS_PADDING 1
 #else
 #    error "no PyrSlot imlementation for this platform"
 #endif
 
-#include <string>
+// On 64-bit systems the pointer is assumed to fit into 48 bits.
+// This is not true for some very modern intel systems which use 56 bits
+//      - but this is not in common use and nan boxing is very common,
+//      so it is likely there will be a workaround (...hopefully...).
+
+//        |-----------------| these are free in a pointer
+// Ptr  = 1111 1111 1111 1111 000000000000000000000000000000000000000000000000
+
+//  Double spec:
+//     sign bit
+//        |  exponent                       fraction
+//        |   11-bit                         52-bit
+//        ||-----------| |---------------------------------------------------|
+//  Inf = 0111 1111 1111 0000 000000000000000000000000000000000000000000000000 = 7FF0 0000 0000 0000
+// -Inf = 1111 1111 1111 0000 000000000000000000000000000000000000000000000000 = FFF0 0000 0000 0000
+
+// If exponents are all 1 and fraction isn't 0, its some type of nan
+// sNaN = 0111 1111 1111 0000 000000000000000000000000000000000000000000000001 = 7FF0 0000 0000 0001
+// qNaN = 0111 1111 1111 1000 000000000000000000000000000000000000000000000001 = 7FF8 0000 0000 0001
+// qNaN = 1111 1111 1111 1000 000000000000000000000000000000000000000000000001 = FFF8 0000 0000 0001
+//        | sign is irrelevant to nan
+// qNaN = 1111 1111 1111 1111 111111111111111111111111111111111111111111111111 = FFFF FFFF FFFF FFFF
+//                     still qNaN
+
+
+// Signaling NaN (sNan) is used to signal some floating point error has occurred.
+// Quiet NaN (sNan) is normal nan.
+
+// So this means the free bits in each are...
+// 0 = used, 1 = free bit
+// qNan = 1000 0000 0000 0111 111111111111111111111111111111111111111111111111
+// Ptr  = 1111 1111 1111 1111 000000000000000000000000000000000000000000000000
+
+// This leaves these 4 bits in which a tag can be stored.
+// left = 1000 0000 0000 0111 000000000000000000000000000000000000000000000000
+
+// condensing this to 4 bits...
+// 0...000 double - the far right bit is called the boxed bit, it is always set if not a double.
+// 0...001 obj
+// 0...011 int
+// 0...101 sym
+// 0...111 char
+// 1...001 nil|false|true  - these are grouped because they have no data, so they can use the remaining bits.
+// 1...011 ptr
+// 1...101 objHdrPtr
+// 1...111 unused
+
+// To decode a slot value:
+//
+//   * Check Nan bits are set and boxed bit*
+//      | false             | true
+//  * double *              |
+//                          |
+//                          |
+//                *
+
+
+static constexpr uint64_t boxedBit = 0x0001000000000000;
+static constexpr uint64_t nanExponent = 0x7FF8000000000000;
+static constexpr uint64_t boxed = boxedBit | nanExponent;
+
+static constexpr uint64_t pointerMask = 0x0000FFFFFFFFFFFF;
+
+static constexpr uint64_t objHdrTag = 0xFFFF000000000000; // boxed | 0x0001000000000000;
```

Comment:
Group these as an enum? e.g.
```
enum class Tag : int {
    objHeader = 0xFFFF000000000000
    // ...etc
};
```

Reply:
I went with 
```
struct Masks {
 static constexpr uint64_t boxed = ...;
};
```
Otherwise you have to do something like `static_cast<std::underlying_type<Masks>>(Masks::boxed)` to get the value out correctly...